### PR TITLE
GH-39024: [Compute] Allow implicitly casting extension to storage types in compute functions

### LINKED ISSUE #39024: [C++] How to use CallFunction() when arg is a ExtensionScalar(ExtensionType)
### Describe the usage question you have. Please include as many useful details as  possible.

I have now inherited and implemented the ExtensionType , which storage type is int16(). 
I want to use the ExtensionType's Array and Scalar to call a custom add function by CallFunction. The custom add function using the built-in Kernel for the two int16(). I also wrote a TypeMatcher to match the Kernel.

Currently, two ExtensionArrays can be added together successfully, but there is an issue when adding an ExtensionArray and an ExtensionScalar; it throws a std::bad_cast error.

After investigating, I found that the parent class of ExtensionScalar is Scalar, but it needs to be cast to PrimitiveScalarBase (whose parent class is also Scalar), which fails.

ExtensionType:
```
class SmallintTypecxs : public ExtensionType
{
public:
	SmallintTypecxs() : ExtensionType(int16()) {}

	std::string extension_name() const override { return "smallintcxs"; }

	bool ExtensionEquals(const ExtensionType &other) const override;

	std::shared_ptr<Array> MakeArray(std::shared_ptr<ArrayData> data) const override;

	Result<std::shared_ptr<DataType>> Deserialize(
		std::shared_ptr<DataType> storage_type,
		const std::string &serialized) const override;

	std::string Serialize() const override { return "smallintcxs"; }

	const std::shared_ptr<DataType>& storage_type() const { return storage_type_; }
private:
};
std::shared_ptr<DataType> smallintcxs() { return std::make_shared<SmallintTypecxs>(); }
```

TypeMatcher
```
//============ type matcher
template <typename ArrowType>
class ExtendMatcher : public TypeMatcher {
  using ThisType = ExtendMatcher<ArrowType>;

 public:
  explicit ExtendMatcher(Type::type oid)
      : type_id_(oid) {}

  bool Matches(const DataType& type) const override {
    if (type.id() != ArrowType::type_id) {
      return false;
    }

	const auto& ex_type = arrow::internal::checked_cast<const ArrowType&>(type);

	return ex_type.storage_type().get()->id() == type_id_;
  }

  bool Equals(const TypeMatcher& other) const override {
    if (this == &other) {
      return true;
    }
    auto casted = dynamic_cast<const ThisType*>(&other);
    if (casted == nullptr) {
      return false;
    }
    return this->type_id_ == casted->type_id_;
  }

  std::string ToString() const override {
    std::stringstream ss;
    ss << ArrowType::type_name() << "(" << type_id_
       << ")";
    return ss.str();
  }

 private:
	Type::type type_id_;
};
```

```
Status RegisterExtensionFunc()
{
	auto registry = arrow::compute::GetFunctionRegistry();
	auto ty = smallintcxs();

	std::shared_ptr<Function> fc ;
	ARROW_ASSIGN_OR_RAISE(fc, registry->GetFunction("add"));

  	std::vector<TypeHolder> args = {int16(), int16()};

	// auto kernel = func.DispatchExact(args);
	const Kernel* kernel;
	ARROW_ASSIGN_OR_RAISE(kernel, arrow::internal::checked_cast<const ScalarFunction&>(*fc)
                             .DispatchExact(args));
	
	const arrow::compute::ScalarKernel* derived_ptr = static_cast<const arrow::compute::ScalarKernel*>(kernel);

	std::cout<<derived_ptr->signature->out_type().ToString()<<std::endl;

        auto func = std::make_shared<ScalarFunction>("cxsadd", Arity::Binary(),
                                                 /*doc=*/FunctionDoc::Empty());

        InputType in_type(SmallIntcxs(ty.get()->storage_id()));

        ARROW_RETURN_NOT_OK(func->AddKernel({in_type,in_type}, int16(), derived_ptr->exec)); 
	
	ARROW_RETURN_NOT_OK(registry->AddFunction(func));

	return Status::OK();
}

```

CallFunction
```
arrow::Status testSmallIntType()
{

	ARROW_RETURN_NOT_OK(arrow::compute::RegisterExtensionFunc());

	auto ary1 = arrow::SmallintArraycxsFromJSON("[30000, 100, 200, 1, 2]");

	auto ary2 = arrow::SmallintArraycxsFromJSON("[30000, 100, 200, 1, 2]");
	
	std::shared_ptr<arrow::Scalar> scalar1;

	ARROW_ASSIGN_OR_RAISE(scalar1, arrow::MakeScalar(arrow::smallintcxs(), 1));

	arrow::Datum incremented_datum;


	//ARROW_ASSIGN_OR_RAISE(incremented_datum,
	//					  arrow::compute::CallFunction("cxsadd", {ary1, ary2})); // it's ok

	ARROW_ASSIGN_OR_RAISE(incremented_datum,
						  arrow::compute::CallFunction("cxsadd", {ary1, scalar1})); // not work
	std::shared_ptr<Array> incremented_array = std::move(incremented_datum).make_array();

	auto resarr = std::static_pointer_cast<arrow::Int16Array>(incremented_array);

	std::cout<<"begin output"<<std::endl;

	for(int i = 0; i < resarr->length(); i++)
	{
		std::cout << static_cast<int64_t>(resarr->Value(i)) << std::endl;
	}
  return arrow::Status::OK();
}
```
When I use 'ARROW_ASSIGN_OR_RAISE(incremented_datum,
						  arrow::compute::CallFunction("cxsadd", {ary1, scalar1})); ', program will abort and show this: 
```
terminate called after throwing an instance of 'std::bad_cast'
  what():  std::bad_cast
```

coredump file:

```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007fd65aa73859 in __GI_abort () at abort.c:79
#2  0x00007fd65ae708d1 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007fd65ae7c37c in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007fd65ae7c3e7 in std::terminate() () from /lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007fd65ae7c699 in __cxa_throw () from /lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007fd65ae7033c in __cxa_bad_cast () from /lib/x86_64-linux-gnu/libstdc++.so.6
#7  0x0000559b33971f3f in arrow::internal::checked_cast<arrow::internal::PrimitiveScalarBase const&, arrow::Scalar const&> (value=...)
    at /home/gpadmin/gitRepo/arrow/cpp/src/arrow/util/checked_cast.h:38
#8  0x0000559b3395fbae in arrow::compute::internal::UnboxScalar<arrow::Int16Type, void>::Unbox (val=...) at /home/gpadmin/gitRepo/arrow/cpp/src/arrow/compute/kernels/codegen_internal.h:341
#9  0x0000559b33cd3331 in arrow::compute::internal::applicator::ScalarBinary<arrow::Int16Type, arrow::Int16Type, arrow::Int16Type, arrow::compute::internal::Add>::ArrayScalar (
    ctx=0x559b37531198, arg0=..., arg1=..., out=0x7fffb3211140) at /home/gpadmin/gitRepo/arrow/cpp/src/arrow/compute/kernels/codegen_internal.h:753
#10 0x0000559b33cc0df8 in arrow::compute::internal::applicator::ScalarBinary<arrow::Int16Type, arrow::Int16Type, arrow::Int16Type, arrow::compute::internal::Add>::Exec (ctx=0x559b37531198,
    batch=..., out=0x7fffb3211140) at /home/gpadmin/gitRepo/arrow/cpp/src/arrow/compute/kernels/codegen_internal.h:780
#11 0x0000559b33686c08 in arrow::compute::detail::(anonymous namespace)::ScalarExecutor::ExecuteSingleSpan (this=0x559b37531090, input=..., out=0x7fffb3211140)
    at /home/gpadmin/gitRepo/arrow/cpp/src/arrow/compute/exec.cc:891
#12 0x0000559b3368665f in arrow::compute::detail::(anonymous namespace)::ScalarExecutor::ExecuteSpans (this=0x559b37531090, listener=0x7fffb3211350)
    at /home/gpadmin/gitRepo/arrow/cpp/src/arrow/compute/exec.cc:859
#13 0x0000559b33686001 in arrow::compute::detail::(anonymous namespace)::ScalarExecutor::Execute (this=0x559b37531090, batch=..., listener=0x7fffb3211350)
    at /home/gpadmin/gitRepo/arrow/cpp/src/arrow/compute/exec.cc:808
#14 0x0000559b336d267d in arrow::compute::detail::FunctionExecutorImpl::Execute (this=0x559b37531170, args=std::vector of length 2, capacity 2 = {...}, passed_length=-1)
    at /home/gpadmin/gitRepo/arrow/cpp/src/arrow/compute/function.cc:276
#15 0x0000559b336d0262 in arrow::compute::(anonymous namespace)::ExecuteInternal (func=..., args=std::vector of length 2, capacity 2 = {...}, passed_length=-1, options=0x0,
    ctx=0x559b359cf320 <arrow::compute::default_exec_context()::default_ctx>) at /home/gpadmin/gitRepo/arrow/cpp/src/arrow/compute/function.cc:341
#16 0x0000559b336d039e in arrow::compute::Function::Execute (this=0x559b37530270, args=std::vector of length 2, capacity 2 = {...}, options=0x0,
    ctx=0x559b359cf320 <arrow::compute::default_exec_context()::default_ctx>) at /home/gpadmin/gitRepo/arrow/cpp/src/arrow/compute/function.cc:348
#17 0x0000559b3368a320 in arrow::compute::CallFunction (func_name="cxsadd", args=std::vector of length 2, capacity 2 = {...}, options=0x0,
    ctx=0x559b359cf320 <arrow::compute::default_exec_context()::default_ctx>) at /home/gpadmin/gitRepo/arrow/cpp/src/arrow/compute/exec.cc:1369
#18 0x0000559b3368a3db in arrow::compute::CallFunction (func_name="cxsadd", args=std::vector of length 2, capacity 2 = {...}, ctx=0x0)
    at /home/gpadmin/gitRepo/arrow/cpp/src/arrow/compute/exec.cc:1374
#19 0x0000559b332837d8 in testSmallIntType () at /home/gpadmin/testProjects/arrow_test/main.cpp:88
#20 0x0000559b33283de8 in main () at /home/gpadmin/testProjects/arrow_test/main.cpp:140
```

### Component(s)

C++

### PR DESCRIPTION
**Pull request by js8544**
### Rationale for this change

Extension types are only logical, i.e. each extension type has a underlying native storage type. We can allow implicit casting from extension types to their storage types. I believe this can make `ExtensionType`s much easier to use.

### What changes are included in this PR?

1. In `DispatchBest`, if no matching kernels exist, allow extension types to be casted to their storage types and match again. The kernel with the minimum number of casts is selected. If there are multiple best kernels, report an error. The dispatching algo is similar to C++'s.
3. In functions that overwrite `DispatchBest`, call `EnsureExtensionToStorage` to simply replace extension types by their storage types, because these functions do not provide native kernels for extension types so we don't need to count the number of casts needed.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

No.

* Closes: #39024

### CODE REVIEW

**Comment by pitrou** [comment]
I am not sure it's a good idea to do this automatically and exhaustively. Some operations will depend on the logical meaning of the types: you probably don't want the same casting rules to apply to `UUID` than to `fixed_size_binary<16>`, for example.

This seems reasonable to functions that merely forward one of their arguments, though, such as `if_else`.

**Comment by js8544** [comment]
I see. I think there are two things we can do do mitigate this problem:
1. Implicit casts should be disabled for Cast functions.
2. Each ExtensionType can have an `allow_implicit_cast_` member to opt-in/opt-out of this feature. This is kind of like the C++ `explicit` keyword: A type needs to opt-out of implicit casting.

**Comment by js8544** [comment]
There's also a more passive yet controlled approach: We add a `std::vector<std::string> allowed_implicit_casting_functions_` member to each ExtensionType and we implicit cast it only for those functions.
It can also be `std::vector<std::string> disallowed_implicit_casting_functions_`.

**Comment by pitrou** [comment]
> There's also a more passive yet controlled approach: We add a `std::vector<std::string> allowed_implicit_casting_functions_` member to each ExtensionType and we implicit cast it only for those functions.

That would work but that means a lot of work for each extension type to follow additions to the range of default compute functions.

I would suggest we perhaps need a more general semantic description of storage type equivalence.
issues?q=is%3Aissue+involves%3Aozankabak+
Draft:
```c++
class ExtensionType {
 public:
  // Storage equivalence for equality testing and hashing
  static constexpr uint32_t kEquality = 1;
  // Storage equivalence for ordered comparisons
  static constexpr uint32_t kOrdering = 2;
  // Storage equivalence for selections (filter, take, etc.)
  static constexpr uint32_t kSelection = 4;
  // Storage equivalence for arithmetic
  static constexpr uint32_t kArithmetic = 8;
  // Storage equivalence for explicit casts
  static constexpr uint32_t kCasting = 16;
  // Storage equivalence for all operations
  static constexpr uint32_t kAny = std::numeric_limits<uint32_t>::max();

  // By default, an extension type can be implicitly handled as its storage type
  // for selections, equality testing and hashing.
  virtual uint32_t storage_equivalence() const { return kEquality | kSelection; }
```

@felipecrv WDYT?

@alamb @tustvold How do you handle extension types in Datafusion?

**Comment by alamb** [comment]
> @alamb @tustvold How do you handle extension types in Datafusion?

TLDR is that we don't currently handle them

We have discussed how to do so in https://github.com/apache/arrow-datafusion/issues/7923 and https://github.com/apache/arrow-rs/issues/4472 but I would say we have not yet reached a consensus

**Comment by bkietz** [comment]
> I would suggest we perhaps need a more general semantic description of storage type equivalence.

Another way to handle this would be with a property on kernels (or even functions) instead of on ExtensionType:

```c++
class ARROW_EXPORT InputType {
 public:
  bool accepts_extension_types_if_storage_matches = false;
  // ...

    // ...
    filter_kernel->signature->in_types()[0].accepts_extension_types_if_storage_matches = true;
```

This would have the effect of disabling operation on extension arrays' storage by default, but allow enabling it per kernel or function. (The filter for accepted extension types could of course also be more articulated than a simple `bool`.)

I think this would put configurability where it needs to be since the specific categories selection, equality comparison, and arithmetic have differing levels of difficulty in support:
- I can't think of a type which would not be selectable by operating on its storage unless the semantics of an array slot depended on its position in the array or on its neighbors somehow, which seems like something incompatible with arrow arrays anyway.
- Equality comparison for extension types would have to be superset of equality comparison for their storage; if the storage slots are identical then whatever the extension type the corresponding slot in the extension array must be identical also. However it would be possible to define an extension of fixed size list wherein the semantic value of a slot is the sum of list elements from the storage's slot, in which case `[1, 2]` would be equal to `[1, 2]` but also equal to `[3, 0]`.
- Arithmetic is much more complex since even if a type supports addition it might not support multiplication and it might only support an addend of a specific other type (as timestamps can be added to durations but not to other timestamps).

Given that supporting various operations on the stored types is so nuanced, it seems we'd need to reserve handling of it for a system which can express those nuances - like kernel dispatch.

---

Another solution would be to formalize what we (de facto) have done thus far: allow casting to/from storage types, allowing operation on storage only when explicitly requested.

**Comment by felipecrv** [comment]
+1 on what @bkietz said: the casting behavior should be declared on the operation, not on the type itself.

Ideally we would have a declarative way of defining more elaborate type constraints on kernels and a decidable solver that dispatches to the appropriate implementation. If we implement that we have to write a checker that ensures there is always only one solution (ordering of declaration of the kernel implementations shouldn't matter).

**Comment by pitrou** [comment]
Ok, say I have a IPv4 extension type backed by a UInt32 storage type. I want to declare that ordering on IPv4 extension values is the same as on the storage type (so that, e.g., sorting would work automatically). How does that work with the `accepts_extension_types_if_storage_matches` proposal?

**Comment by felipecrv** [comment]
> Ok, say I have a IPv4 extension type backed by a UInt32 storage type. I want to declare that ordering on IPv4 extension values is the same as on the storage type (so that, e.g., sorting would work automatically). How does that work with the `accepts_extension_types_if_storage_matches` proposal?

The operation declares that it can receive int32 and all logical types defined in terms of int32 (like `IPv4`).

`sort(array<int32>)` can also work as `sort(array<ipv4>)`, but `sum(int32, int32)` can't.

But to be fair, I would have to think more carefully about these things before accepting any magical implicit behavior. See how much pain the C implicit casts (that look inoffensive) have created in the world.

**Comment by pitrou** [comment]
> The operation declares that it can receive int32 and all logical types defined in terms of int32 (like `IPv4`).
> 
> `sort(array<int32>)` can also work as `sort(array<ipv4>)`, but `sum(int32, int32)` can't.

Yes, but the problem here is another extension type based on  `uint32` would also inherit the sort behavior even though it might want a different definition of ordering.

**Comment by tustvold** [comment]
FWIW the design in arrow-rs currently is that extension types *only* exist in the metadata. The rationale being that by default kernels should treat them as the storage type. Imo it goes against the whole purpose of extension types if the only way to support them is for every kernel to be aware of all supported extension types.

That's not to say there can't be special kernels for certain extension types, potentially selected as part of a query engine's planning, but I can't help feeling if an array can't be correctly treated as its physical type, something is a bit off.

That's all to say I agree with coercing extension arrays to their physical types, arrow-rs goes even further by not having an extension array at all. 

That all being said, that is just my view on this, and I suspect others will feel differently

**Comment by yukkit** [comment]
Mark, to be addressed later

**Comment by bkietz** [comment]
I think this conversation needs to move to the mailing list. I'll open a thread with a summary

**Comment by bkietz** [comment]
https://lists.apache.org/thread/2j61nrod7x0s5vjhc6q9tlj898drz7rn

**Comment by mapleFU** [comment]
I think another problem it's that maybe it's hard to define some logic in extension type without or even with these casts.

For example, if I want to define a "cast_extension_to_string", I cannot add my logic to the MetaFunction "Cast", do we have some solving for this?

**Comment by thisisnic** [comment]
Thank you for your contribution. Unfortunately, this 
  pull request has been marked as stale because it has had no activity in the past 365 days. Please remove the stale label 
  or comment below, or this PR will be closed in 14 days.  Feel free to re-open this if it has been closed in error. If you
   do not have repository permissions to reopen the PR, please tag a maintainer.
